### PR TITLE
Issue 802

### LINKED ISSUE #802: Accuracy computation with variable classes
## ❓ Questions/Help/Support

I've been using the `Accuracy` class (in `ignite.metrics`) successfully so far in a multi-class setting to compute test set accuracy.

Recently, I've hit a new scenario/dataset where the test set has different number of classes for each example. You can think of this as a multiple-choice selection task, where each example has a different number of candidates.

Now in this scenario, when I used `Accuracy`, I got the below error:
`ERROR:ignite.engine.engine.Engine:Current run is terminating due to exception: Input data number of classes has changed from 13 to 81.`

I dived into the [code](https://github.com/pytorch/ignite/blob/master/ignite/metrics/accuracy.py#L83L92) and I see that the base class for `Accuracy`, i.e., `_BaseClassification ` assumes that the number of classes/candidates is fixed across test set examples. And this is why I'm getting the above error at run-time.

However, for multi-class accuracy, as you can see in the `update()` method [here](https://github.com/pytorch/ignite/blob/master/ignite/metrics/accuracy.py#L154L156), we are simply computing the number of correct predictions via `argmax` along dimension-1 and measuring matches against `y`.

This computation of `correct` should be accurate even if the number of candidates changes across two examples from the test set, right?

So in effect, the computed accuracy would be correct even in such a case because the underlying variables are computed [correctly](https://github.com/pytorch/ignite/blob/master/ignite/metrics/accuracy.py#L165L166)?

What do you think, @vfdev-5? Is it possible to have a version of `Accuracy` for multi-class where there is no expectation that the number of classes is the same across all examples in the test set?

### PR DESCRIPTION
**Pull request by ykumards**
Fixes #802 

Description: Add flag to `Accuracy` to handle computation with varying number of classes in multi-class setting

Check list:
* [x] New tests are added (if a new feature is added)
* [ ] New doc strings: description and/or example code are in RST format
* [x] Documentation is updated (if required)

### CODE REVIEW

**Comment by ykumards** [comment]
Hi @vfdev-5 does this look like what we'd discussed about #802? Also, am not very confident about the name of the flag I've used here, do you have a better suggestion?

File: ignite/metrics/accuracy.py [hunk 1]
```diff
@@ -14,9 +14,11 @@ def __init__(
         self,
         output_transform: Callable = lambda x: x,
         is_multilabel: bool = False,
+        is_variable_multiclass: bool = False,
```
Comment (justusschock) [inline ignite/metrics/accuracy.py]: can't we rename this to something like `variable_classes`? We already have `is_multilabel` here and together they should be the same,  but maybe we can reuse `variable_classes` in some further way.

So `is_variable_multiclass` would be the same as `is_multilabel and variable_classes`. This won't change the functionality though but we don't have separate variables indicating the same thing (partially)

Comment (vfdev-5) [reply ignite/metrics/accuracy.py]: `variable_classes` sounds good. However, we need emphasize that this is possible only for multiclass case and not multi-label.

File: ignite/metrics/accuracy.py [hunk 2]
```diff
@@ -14,9 +14,11 @@ def __init__(
...
         device: Optional[Union[str, torch.device]] = None,
     ):
         self._is_multilabel = is_multilabel
+        self._is_variable_multiclass = is_variable_multiclass
```
Comment (vfdev-5) [inline ignite/metrics/accuracy.py]: We need to check that this flag is False in case of multi-label

File: ignite/metrics/accuracy.py (hunk 1 above)
Comment (ykumards) [reply ignite/metrics/accuracy.py]: sounds good. I pushed the changes

File: ignite/metrics/accuracy.py (hunk 2 above)
Comment (ykumards) [reply ignite/metrics/accuracy.py]: done!

**Comment by sdesrozis** [comment]
@ykumards @vfdev-5

Does this PR need more work or could we merge / close it?

**Comment by vfdev-5** [comment]
@sdesrozis we wanted to refactor these classes in case of multi-label. So this code is a temporary solution to fix the issue.

**Comment by vfdev-5** [comment]
Well, let's merge and we refactor thing later.

**Comment by sdesrozis** [comment]
It seems that the branch of @ykumards is not longer reachable.

**Comment by Moh-Yakoub** [comment]
@sdesrozis  @vfdev-5 It seems that this PR went inactive for a while, if we are still interested in that change I can prepare a similar PR on a more fresh master branch.

**Comment by sdesrozis** [comment]
I let @vfdev-5 choose what to do. Maybe @g-karthic would tell us if this PR is still relevant.

**Comment by vfdev-5** [comment]
@Moh-Yakoub @sdesrozis let's close this PR as it would only represent a temporary solution. Let's discuss more in the related issue.

**Pull request closed without merging**